### PR TITLE
[aarch64] fix page table flags; add identity mapping of device MMIO

### PR DESCRIPTION
**Pull request by shrik3**
- HYPERCALL_MMIO_BASE is relocated to 2 pages below PHY_LOW_ADDR  (256G - 8K)
- It turns out that with MMU we don't need to align (mmio) hypercall addresses to 8 bytes.

test run from our arm machine (Neoverse-N1). I annotated the hypercalls. 

```
tack is 45c0080000
[INFO] [0/225894469457] enter vcpu run
[INFO] [0/225894469459] kvmRet: MmioWrite(274877898760, [0, 0, 0, 0])
[INFO] [0/225894469459] handle hypercall 274877898760    <- HYPERCALL_PRINT
[INFO] [0/225894469459] paras:45c007fdd0 0 0 0
0x40003d622f 0x45c007fde0: enter rust_main, haha         <- rust_main
[INFO] [0/225894469459] enter vcpu run
[INFO] [0/225894469460] kvmRet: MmioWrite(274877898760, [0, 0, 0, 0])
[INFO] [0/225894469460] handle hypercall 274877898760    <- HYPERCALL_PRINT
[INFO] [0/225894469460] paras:45c007fc68 0 0 0
0x40003d6244 0x45c007fc78: init global allocator   
[INFO] [0/225894469460] enter vcpu run
[INFO] [0/225894469460] kvmRet: MmioWrite(274877898760, [0, 0, 0, 0])
[INFO] [0/225894469460] handle hypercall 274877898760    <- HYPERCALL_PRINT
[INFO] [0/225894469460] paras:45c007fc68 0 0 0
0x40003d6259 0x45c007fc78: init global allocator2        <- rust_main
[INFO] [0/225894469460] enter vcpu run
[INFO] [0/225894469461] kvmRet: MmioWrite(274877898757, [0, 0, 0, 0])
[INFO] [0/225894469461] handle hypercall 274877898757    <- HYPERCALL_MSG
[INFO] [0/225894469461] paras:240 0 3f 2
[INFO] [0/225894469461] enter vcpu run
[INFO] [0/225894469462] kvmRet: MmioWrite(274877898760, [0, 0, 0, 0])
[INFO] [0/225894469462] handle hypercall 274877898760    <- HYPERCALL_PRINT
[INFO] [0/225894469462] paras:45c007fc68 0 0 0
0x40003d626f 0x45c007fc78: init sharespace               <- rust_main
[INFO] [0/225894469462] enter vcpu run
[INFO] [0/225894469462] kvmRet: MmioWrite(274877898757, [0, 0, 0, 0])
[INFO] [0/225894469462] handle hypercall 274877898757    <- HYPERCALL_MSG
[INFO] [0/225894469462] paras:240 0 3f 3                 <- task.rs:154
[INFO] [0/225894469463] enter vcpu run                                  
[INFO] [0/225894469466] kvmRet: MmioWrite(274877898757, [0, 0, 0, 0])
[INFO] [0/225894469466] handle hypercall 274877898757    <- HYPERCALL_MSG
[INFO] [0/225894469466] paras:237 0 0 0                  <- task.rs:383
[INFO] [0/225894469466] enter vcpu run
[INFO] [0/225894469467] kvmRet: MmioWrite(274877898773, [0, 0, 0, 0])
[INFO] [0/225894469467] handle hypercall 274877898773    <- HYPERCALL_VCPU_DEBUG
[INFO] [0/225894469467] paras:0 0 0 0
[ERROR] [0/225894469467] HYPERCALL_VCPU_DEBUG
[INFO] [0/225894469467] enter vcpu run
                                                         <- HANGS HERE
```

### CODE REVIEW

File: qvisor/src/runc/runtime/vm.rs [hunk 1]
```diff
@@ -349,13 +349,17 @@ impl VirtualMachine {
...
+ 
+            let mut opts = addr::PageOpts::Zero();
 
-            // vms.KernelMap(
-            //     addr::Addr(MemoryDef::HYPERCALL_MMIO_BASE),
-            //     addr::Addr(MemoryDef::HYPERCALL_MMIO_BASE + 0x1000),
-            //     addr::Addr(MemoryDef::HYPERCALL_MMIO_BASE),
-            //     opts.Val(),
-            // )?;
+			opts.SetWrite().SetGlobal().SetPresent().SetAccessed();
```
Comment (shrik3) [inline qvisor/src/runc/runtime/vm.rs:355]: it seems that I mixed several tab indentations here :clown_face:  .  We can polish this in future commits ..

File: qlib/kernel/arch/aarch64/mm/pagetable.rs [hunk 2]
```diff
@@ -119,9 +120,13 @@ bitflags! {
...
         const VALID           = 1;
         const TABLE           = 1 << 1;
-        const USER_ACCESSIBLE = 1 << 6;
+        const PAGE            = 3;      // 4k granule
+        const DEVICE          = 1 << 2; // device mmio
+        const NC              = 1 << 3; // non cachable
+        const USER_ACCESSIBLE = 1 << 6; // allow EL0 access
         const READ_ONLY       = 1 << 7;
-        const SHARED          = 3 << 8;
+        const OUTTER_SHARABLE = 2 << 8;
```
Comment (SchwarzStein) [inline qlib/kernel/arch/aarch64/mm/pagetable.rs:128]: fix the misspelling

File: qlib/kernel/arch/aarch64/mm/pagetable.rs [hunk 3]
```diff
@@ -119,9 +120,13 @@ bitflags! {
     pub struct PageTableFlags: u64 {
         const VALID           = 1;
         const TABLE           = 1 << 1;
-        const USER_ACCESSIBLE = 1 << 6;
+        const PAGE            = 3;      // 4k granule
+        const DEVICE          = 1 << 2; // device mmio
```
Comment (SchwarzStein) [inline qlib/kernel/arch/aarch64/mm/pagetable.rs:124]: This flag DEVICE is same as MT_DEVICE_NGNRE below.
remove duplicates.
I suggest u rename the two to: MEM_DEVICE and MEM_NORMAL, for device memory and normal memory respectively.

File: qlib/kernel/arch/aarch64/mm/pagetable.rs [hunk 4]
```diff
@@ -132,8 +137,16 @@ bitflags! {
         const ZERO            = 0;
         const PRESENT         = 1;
 
-        const MT_DEGICE_NGNRE = 0x1 << 2;
+        const MT_DEVICE_NGNRE = 0x1 << 2;
         const MT_NORMAL       = 0x4 << 2;
+        // A handy collection of flags that specify device memory
+        const DEVICE_FLAGS    = Self::DEVICE.bits|
+                                Self::UXN.bits|
```
Comment (SchwarzStein) [inline qlib/kernel/arch/aarch64/mm/pagetable.rs:144]: add the PXN flag and test to confirm ir works,

File: qlib/kernel/arch/aarch64/mm/pagetable.rs [hunk 5]
```diff
@@ -119,9 +120,13 @@ bitflags! {
     pub struct PageTableFlags: u64 {
         const VALID           = 1;
```
Comment (SchwarzStein) [inline qlib/kernel/arch/aarch64/mm/pagetable.rs:121]: could you please add the other flags: PXN , ....
would be nice if they followed the manual for the sake of completeness.

if these flags are exclusive to aarch64 and not shared with x86 can we prefix with PT_?

**Review by SchwarzStein — COMMENTED** [review]
please see the requested changes

File: qlib/kernel/arch/aarch64/mm/pagetable.rs (hunk 3 above)
Comment (shrik3) [reply qlib/kernel/arch/aarch64/mm/pagetable.rs:124]: (TODO/suggestion) multiple-bit flags should be avoided for 
some fields e.g. Memory Attributes see  
https://docs.rs/bitflags/latest/bitflags/#multi-bit-flags

File: qlib/kernel/arch/aarch64/mm/pagetable.rs (hunk 5 above)
Comment (shrik3) [reply qlib/kernel/arch/aarch64/mm/pagetable.rs:121]: > if these flags are exclusive to aarch64, and not shared with x86, could we prefix them with PT_?

Refactoring those names would be beyond the scope of this PR. We can resolve this in another issue.

File: qlib/kernel/arch/aarch64/mm/pagetable.rs (hunk 4 above)
Comment (shrik3) [reply qlib/kernel/arch/aarch64/mm/pagetable.rs]: Added. and it works (on my machine)

File: qlib/kernel/arch/aarch64/mm/pagetable.rs (hunk 3 above)
Comment (SchwarzStein) [reply qlib/kernel/arch/aarch64/mm/pagetable.rs]: That's the meaning of each of those flag. It does not imply multiple bits

**Pull request merged**